src/host-config/redsocks.ts

```diff
@@ -0,0 +1,352 @@
+
+	// Get and parse no_proxy
+	const noProxy = await readNoProxy();
+
+	// Build proxy object
+	const proxy = {
+		...HostConfig.fromRedsocksConfig(redsocksConf),
+		...(noProxy.length && { noProxy }),
+	};
+
+	// Remove double quotes from auth fields
+	if (proxy.login) {
+		proxy.login = proxy.login.replace(/"/g, '');
+	}
+	if (proxy.password) {
+		proxy.password = proxy.password.replace(/"/g, '');
+	}
+
+	// Convert port to number per API doc spec
+	if (proxy.port) {
+		proxy.port = parseInt(proxy.port as any, 10);
+	}
+
+	// Assumes mandatory proxy config fields (type, ip, port) are present,
+	// even if they very well may not be. It is up to the user to ensure
+	// that all the necessary fields are present in the redsocks.conf file.
+	return proxy as HostProxyConfig;
+}
+
+export async function setProxy(conf: HostProxyConfig) {
+	// Ensure proxy directory exists
+	await mkdirp(proxyBasePath);
+
+	// Set no_proxy
+	const noProxy = conf.noProxy;
+	const noProxyChanged = await setNoProxy(noProxy);
+
+	// Remove noProxy from target redsocks.conf
+	const targetConf = Object.fromEntries(
+		Object.entries(conf)
+			// Remove auxiliary fields
+			.filter(([key]) => !['noProxy'].includes(key)),
+	) as WriteableProxyConfig;
+
+	// If target is empty aside from noProxy and noProxy got patched,
+	// do not change redsocks.conf to remain backwards compatible
+	if (Object.keys(targetConf).length === 0 && noProxyChanged) {
+		// Restart services using dbus
+		await restartProxyServices();
+		return;
+	}
+
+	// Set redsocks.conf
+	const currentHostConf = await readProxy();
+	const currentConf = Object.fromEntries(
+		Object.entries(currentHostConf ?? {})
+			// Remove auxiliary fields
+			.filter(([key]) => !['noProxy'].includes(key)),
+	) as WriteableProxyConfig;
+
+	// Merge current & target redsocks.conf
+	const patchedConf = RedsocksConf.patch(
+		HostConfig.toRedsocksConfig(currentConf),
+		HostConfig.toRedsocksConfig(targetConf),
+	);
+
+	// Write to redsocks.conf
+	const toWrite = RedsocksConf.stringify(patchedConf);
+	if (toWrite) {
+		await writeToBoot(redsocksConfPath, toWrite);
+	} else {
+		await unlinkAll(redsocksConfPath);
+	}
+
+	// Restart services using dbus
+	await restartProxyServices();
+}
+
+async function restartProxyServices() {
+	// restart balena-proxy-config if it is loaded and NOT PartOf redsocks-conf.target
+	if (
+		(
+			await Promise.any([
+				dbus.servicePartOf('balena-proxy-config'),
+				dbus.servicePartOf('resin-proxy-config'),
+			])
+		).includes('redsocks-conf.target') === false
+	) {
+		await Promise.any([
+			dbus.restartService('balena-proxy-config'),
+			dbus.restartService('resin-proxy-config'),
+		]);
+	}
+
+	// restart redsocks if it is loaded and NOT PartOf redsocks-conf.target
+	if (
+		(await dbus.servicePartOf('redsocks')).includes('redsocks-conf.target') ===
+		false
+	) {
+		await dbus.restartService('redsocks');
+	}
+}
+
+export async function readNoProxy(): Promise<string[]> {
```

Comment:
I'm divided on whether the noproxy code belongs here, perhaps the module would have to be called `proxy.ts`

Reply:
IMO no_proxy is a layer of configuration applied by the host OS on top of redsocks, so renaming to `proxy.ts` makes sense, or even having `noProxy.ts` as a separate file